File: 1638.count-substrings-that-differ-by-one-character.go
// @leet start
func countSubstrings(s string, t string) int {
	f := make([][]int, len(s)+1)
	for i := 0; i < len(f); i++ {
		f[i] = make([]int, len(t)+1)
	}
	for i := 0; i < len(s); i++ {
		for j := 0; j < len(t); j++ {
			if s[i] == t[j] {
				f[i+1][j+1] = f[i][j] + 1
			}
		}
	}
	g := make([][]int, len(s)+1)
	for i := 0; i < len(g); i++ {
		g[i] = make([]int, len(t)+1)
	}
	res := 0
	for i := len(s) - 1; i >= 0; i-- {
		for j := len(t) - 1; j >= 0; j-- {
			if s[i] == t[j] {
				g[i][j] = g[i+1][j+1] + 1
			} else {
				res += (f[i][j] + 1) * (g[i+1][j+1] + 1)
			}
		}
	}
	return res
}

// @leet end

File: 1638.count-substrings-that-differ-by-one-character.cs
// @leet start
public class Solution {
    public int CountSubstrings(string s, string t) {
      var f = new int[s.Length + 1, t.Length + 1];
      for (int i = 0; i < s.Length; i++) {
        for (int j = 0; j < t.Length; j++) {
          if (s[i] == t[j]) {
            f[i + 1, j + 1] = f[i, j] + 1;
          }
        }
      }
      var g = new int[s.Length + 1, t.Length + 1];
      var res = 0;
      for (int i = s.Length - 1; i >= 0; i--) {
        for (int j = t.Length - 1; j >= 0; j--) {
          if (s[i] == t[j]) {
            g[i, j] = g[i + 1, j + 1] + 1;
          }
          else {
            res += (f[i, j] + 1) * (g[i + 1, j + 1] + 1);
          }
        }
      }
      return res;
    }
}
// @leet end
